### PR TITLE
fixed: fix multitrack support for audio decoders

### PR DESCRIPTION
**Pull request by notspiff**
Somebody has confused tracks and tags, as well as enforced that tags must be supported for tracks to work. Fix this.

### CODE REVIEW

File: xbmc/addons/AudioDecoder.cpp [hunk 1]
```diff
@@ -116,10 +117,18 @@ int CAudioDecoder::GetTrackCount(const std::string& strPath)
...
   int result = m_struct.toAddon.track_count(&m_struct, strPath.c_str());
 
-  if (result > 1 && !Load(strPath, XFILE::CMusicFileDirectory::m_tag, nullptr))
-    return 0;
+  if (result > 1)
+  {
+    if (m_hasTags)
+    {
+      if (!Load(strPath, XFILE::CMusicFileDirectory::m_tag, nullptr))
+      return 0;
```
Comment (xhaggi) [inline xbmc/addons/AudioDecoder.cpp]: wrong indentation

**Comment by notspiff** [comment]
fixed.

**Comment by xhaggi** [comment]
Thx

**Comment by notspiff** [comment]
appended some more fixes;
1) let the file extension provider handle file folder extensions
2) a stupid infinite loop due to the fact that add-on id matches a file folder extension. this leads to an infinite loop (e.g. audiodecoder.wsr and .wsr files)

**Comment by notspiff** [comment]
Shall i invent a fake TYPE entry then? Or make the map a vector of strings? Neither appealed to me..

**Comment by notspiff** [comment]
As in i need to encode the bool in addition to the type. Existence of an addon is not enough

**Comment by xhaggi** [comment]
Sure I meant you should store it in condition of the boolean value into a new string or vector of string member, `m_addonFileFolderExtension` and return it in the getter.

**Comment by xhaggi** [comment]
something like

```c++
for (const auto& addonInfo : addonInfos)
{
  std::string ext = addonInfo->Type(type)->GetValue(ADDON_VFS == type ? "@extensions" : "@extension").asString();
  if (!ext.empty())
  {
    extensions.push_back(ext);

    std::string info = ADDON_VFS == type ? "@filedirectories" : "@tracks";
    if (addonInfo->Type(type)->GetValue(info).asBoolean())
      fileFolderExtensions.push_back(ext);
  }
}

m_addonFileFolderExtensions = StringUtils::Join(fileFolderExtensions, "|");
```

**Comment by notspiff** [comment]
ah, i remember why i didn't do that. because of the the event, it only updates for a single type.

**Comment by xhaggi** [comment]
We could use the same map structure and combine each type in the getter

**Comment by notspiff** [comment]
right. if you think it's worth it, i will do another map.

**Comment by notspiff** [comment]
now with more map.

**Comment by xhaggi** [comment]
Yes please, we don't update the addons very often at runtime and if it's done it will only effect the related types.

**Comment by notspiff** [comment]
sure i meant in terms of code effort ;)

File: xbmc/utils/FileExtensionProvider.cpp [hunk 2]
```diff
@@ -123,6 +145,28 @@ void CFileExtensionProvider::SetAddonExtensions(const TYPE& type)
   m_addonExtensions.insert(make_pair(type, StringUtils::Join(extensions, "|")));
 }
 
+void CFileExtensionProvider::SetAddonFileFolderExtensions(const TYPE& type)
```
Comment (xhaggi) [inline xbmc/utils/FileExtensionProvider.cpp]: Thought we could combine both in `SetAddonExtensions` and add the conditions there, but its fine too.

Comment (notspiff) [reply xbmc/utils/FileExtensionProvider.cpp]: sure. different rules in different projects, my colleagues would hate on me for doing two things in one method.

File: xbmc/utils/FileExtensionProvider.cpp [hunk 3]
```diff
@@ -111,16 +132,30 @@ void CFileExtensionProvider::SetAddonExtensions()
...
+    std::string info = ADDON_VFS == type ? "@extensions" : "@extension";
+    std::string ext = addonInfo->Type(type)->GetValue(info).asString();
     if (!ext.empty())
       extensions.push_back(ext);
+    if (type == ADDON_VFS || type == ADDON_AUDIODECODER)
+    {
+      std::string info2 = ADDON_VFS == type ? "@filedirectories" : "@tracks";
+      if (addonInfo->Type(type)->GetValue(info2).asBoolean())
+      {
+        ext = addonInfo->Type(type)->GetValue(info).asString();
```
Comment (xhaggi) [inline xbmc/utils/FileExtensionProvider.cpp]: No need to get this value again, it's already done on line 141

File: xbmc/utils/FileExtensionProvider.cpp [hunk 4]
```diff
@@ -111,16 +132,30 @@ void CFileExtensionProvider::SetAddonExtensions()
...
   BinaryAddonBaseList addonInfos;
   CServiceBroker::GetBinaryAddonManager().GetAddonInfos(addonInfos, true, type);
   for (const auto& addonInfo : addonInfos)
   {
-    std::string ext = addonInfo->Type(type)->GetValue(ADDON_VFS == type ? "@extensions" : "@extension").asString();
+    std::string info = ADDON_VFS == type ? "@extensions" : "@extension";
+    std::string ext = addonInfo->Type(type)->GetValue(info).asString();
     if (!ext.empty())
       extensions.push_back(ext);
+    if (type == ADDON_VFS || type == ADDON_AUDIODECODER)
```
Comment (xhaggi) [inline xbmc/utils/FileExtensionProvider.cpp]: Enclose the code below into `if (!ext.empty())`, we can skip it if ext is empty.

**Review by xhaggi — CHANGES_REQUESTED** [review]

**Comment by notspiff** [comment]
updated.

**Comment by Rechi** [comment]
jenkins build this please

**Comment by xhaggi** [comment]
@notspiff thx

**Review by xhaggi — APPROVED** [review]

**Comment by notspiff** [comment]
approved for a week, no conflicts, is a bug fix. what's holding this back?

**Comment by xhaggi** [comment]
Merge it in

**Pull request merged**